Fix/ds missing distance reserve meal

dsmith42 opened the pull request:
Found a couple of bugs when working on the android app.

When we reserve a meal there was no distance calculation being carried out so the meal returned ended up having null for distance.

~~Changes the recyclerAdapter to use a mutable list so we can update one meal item at a time (e.g. on reserving).~~

Makes the recyclerAdapter use ListAdapter and DiffUtil to update only the items that have changed.  Useful here for when we want to reserve a meal.

dsmith42 committed:
Fixes a bug causing distance not to be calculated

dsmith42 committed:
Merge branch 'main' into fix/ds_missing-distance-reserve-meal

dsmith42 commented on feature-find-meal/src/main/java/com/kinandcarta/feature/find/meal/adapter/MealsAdapter.kt:
```diff
@@ -14,7 +14,7 @@ import com.kinandcarta.feature.find.meal.R
 
 class MealsAdapter(
-    private var mealsList: List<Meal> = listOf(),
+    private var mealsList: MutableList<Meal> = mutableListOf(),
```
Made it mutable so that we can update the item that changed on the reserve only, rather than having to refresh the whole list.  Which is a worse UX.

KaneCheshire replied:
Nice. I think we could also use [ListAdapter](https://developer.android.com/reference/androidx/recyclerview/widget/ListAdapter) to handle a lot of this for us but this works great as well

Gmiller290488 replied:
Playing the other side here:
I'm a user that tried to reserve a meal - turns out my list is not up to date and that meal is gone so I get an alert (or toast) saying "Sorry that meal is unavailable"
and that item is updated to say as much.
I go to the next item nearby and tap to reserve, but this item is also not up to date.
I get the same alert and that one item updates to show it's unavailable.

Could be annoying?

Gmiller290488 replied:
Another thing maybe to consider is that if the user is moving.
One items distance gets updated but the rest of the list doesn't

dsmith42 replied:
Perhaps we need to consider how up to date the list is going to need to be to cover those scenarios?

Gmiller290488 replied:
Not sure what you mean

dsmith42 replied:
What I meant was, once we have a list, after a little bit of time it may be out of date, regarding the reservation status and distance.   So perhaps we might want to think about adding a refresh list reminder or timed / event driven refresh

Gmiller290488 replied:
Yeah something to think on

dsmith42 replied:
Have updated to use ListAdapter diffing, allowing the List to be updated as a whole.

dsmith42 commented on feature-find-meal/src/main/java/com/kinandcarta/feature/find/meal/view/FindMealFragment.kt:
```diff
@@ -92,13 +92,14 @@ class FindMealFragment : Fragment() {
     }
 
     private fun onReservedMeal(state: FindMealViewModel.State.ReservedMeal) {
-        mealsAdapter.notifyItemChanged(state.position)
```
Realised this wasn't working as I had anticipated.  Due to no update actually being carried out on the list.

KaneCheshire replied:
I think what you've changed it to is cleaner anyway, it would just be nice not to have to worry about the position at all and maybe let the adapter figure that out (which I think ListAdapter does). Not a problem for now though, since this works

dsmith42 replied:
Have converted over to use the ListAdapter 👍

dsmith42 commented on feature-find-meal/src/main/java/com/kinandcarta/feature/find/meal/viewmodel/FindMealViewModel.kt:
```diff
@@ -46,6 +46,7 @@ class FindMealViewModel @ViewModelInject constructor(
             updateMeals()
         }
     }
+    private val distanceUnit = DistanceUnit.miles
```
Lifted this up as we use it in 2 places

dsmith42 commented on shared/src/commonMain/kotlin/com/kcc/kmmhackathon/shared/MealsSDK.kt:
```diff
@@ -45,15 +45,24 @@ class MealsSDK {
+                distanceUnit
+            )
+            return updatedMeal
         }
```
Adds the distance calculation to the ReserveMeal in SDK, which is needed here to avoid retuning the Meal with a null distance value.

Gmiller290488 commented on feature-find-meal/src/main/java/com/kinandcarta/feature/find/meal/adapter/MealsAdapter.kt:
```diff
@@ -39,6 +39,11 @@ class MealsAdapter(
     }
 
+    fun updateMeal(position: Int, meal: Meal) {
+        mealsList[position] = meal
+        notifyItemChanged(position)
```
Could you not update just the quantity here, as that is what has changed, and then you wouldn't need the distance code for the reserve call?

dsmith42 replied:
That would do the trick and mean nothing else needs to change on API side.  Well spotted!

dsmith42 committed:
Updates only the quantity value of the item

dsmith42 committed:
Tidying up

dsmith42 committed:
Adds human readable intervals to locationRequest

dsmith42 committed:
Merge branch 'main' into fix/ds_missing-distance-reserve-meal

dsmith42 committed:
Updates Adapter to be a ListAdapter and diffUtil

Improves the implementation by using DiffUtil to detect changes and updates all meals on making a reservation.

dsmith42 committed:
Merge branch 'main' into fix/ds_missing-distance-reserve-meal

# Conflicts:
#	feature-find-meal/src/main/java/com/kinandcarta/feature/find/meal/adapter/MealsAdapter.kt

dsmith42 committed:
Fixes alignment of portions string
